### PR TITLE
Minor Versioning Changes

### PR DESCRIPTION
**Pull request by theunrepentantgeek**
Minor changes to versioning

* Added sorting for slices of `PackageVersion` 
* APIs will be generated with version numbers like `v1alpha1api20180801`
* Added `IsPreview()` to the `PackageReference` interface, implemented and tested
* Updated receiver names to fit usual Go convention

The `IsPreview()` method will be needed to ensure that a preview API version isn't nominated as the official hub storage version.

### CODE REVIEW

**Review by babbageclunk — APPROVED** [review]
Looks great! Just some minor comments and questions.

File: hack/generated/_apis/microsoft.resources/v1alpha1api20200601/groupversion_info_gen.go [hunk 1]
```diff
@@ -3,13 +3,13 @@ Copyright (c) Microsoft Corporation.
 Licensed under the MIT license.
 */
 
-// Package v20200601 contains API Schema definitions for the microsoft.resources v20200601 API group
+// Package v1alpha1api20200601 contains API Schema definitions for the microsoft.resources v20200601 API group
```
[the remark this answers is not in the document]
Comment (theunrepentantgeek) [reply hack/generated/_apis/microsoft.resources/v1alpha1api20200601/groupversion_info_gen.go:6]: Done.

File: hack/generator/pkg/astmodel/external_package_reference.go [hunk 2]
```diff
@@ -25,31 +25,36 @@ func MakeExternalPackageReference(packagePath string) ExternalPackageReference {
 
 // AsLocalPackage returns an empty local reference and false to indicate that library packages
 // are not local
-func (pr ExternalPackageReference) AsLocalPackage() (LocalPackageReference, bool) {
+func (e ExternalPackageReference) AsLocalPackage() (LocalPackageReference, bool) {
```
[the remark this answers is not in the document]
Comment (theunrepentantgeek) [reply hack/generator/pkg/astmodel/external_package_reference.go:28]: As discussed offline, will revert these renames

File: hack/generator/pkg/astmodel/local_package_reference.go [hunk 3]
```diff
@@ -23,49 +26,57 @@ func MakeLocalPackageReference(prefix string, group string, version string) Loca
 }
 
 // AsLocalPackage returns this instance and true
-func (pr LocalPackageReference) AsLocalPackage() (LocalPackageReference, bool) {
-	return pr, true
+func (l LocalPackageReference) AsLocalPackage() (LocalPackageReference, bool) {
```
[the remark this answers is not in the document]
Comment (theunrepentantgeek) [reply hack/generator/pkg/astmodel/local_package_reference.go:29]: As discussed offline, will revert these renames

File: hack/generator/pkg/astmodel/package_reference.go [hunk 4]
```diff
@@ -5,7 +5,11 @@
 
 package astmodel
 
-import "github.com/pkg/errors"
+import (
+	"github.com/pkg/errors"
```
[the remark this answers is not in the document]
Comment (theunrepentantgeek) [reply hack/generator/pkg/astmodel/package_reference.go:9]: Done.

File: hack/generator/pkg/astmodel/package_reference.go [hunk 5]
```diff
@@ -26,6 +30,8 @@ type PackageReference interface {
 	Equals(ref PackageReference) bool
 	// String returns the string representation of the package reference
 	String() string
+	// IsPreview returns true if this package reference is a preview
```
[the remark this answers is not in the document]
Comment (theunrepentantgeek) [reply hack/generator/pkg/astmodel/package_reference.go:33]: Improved anyway.

File: hack/generator/pkg/astmodel/package_reference.go [hunk 6]
```diff
@@ -36,3 +42,219 @@ func PackageAsLocalPackage(pkg PackageReference) (LocalPackageReference, error)
...
+			compare := v.compareNumeric()
+			if compare != 0 {
+				return compare
+			}
+
+			continue
+		}
+
+		if unicode.IsLetter(v.left[v.leftIndex]) && unicode.IsLetter(v.right[v.rightIndex]) {
+			// Found the start of a number
```
[the remark this answers is not in the document]
Comment (theunrepentantgeek) [reply hack/generator/pkg/astmodel/package_reference.go:103]: Fixed.

File: hack/generator/pkg/astmodel/package_reference.go [hunk 7]
```diff
@@ -36,3 +42,219 @@ func PackageAsLocalPackage(pkg PackageReference) (LocalPackageReference, error)
...
+		}
+
+	}
+
+	return 0
+}
+
+// compareNumeric compares two digit sequences as though they represent an integer number. We don't
+// convert the number to a literal int because we don't want to run the risk of overflow, but
+// fortunately we can compare numbers digit by digit.
```
[the remark this answers is not in the document]
Comment (theunrepentantgeek) [reply hack/generator/pkg/astmodel/package_reference.go:133]: Improved the comment.

File: hack/generator/pkg/astmodel/package_import_set.go [hunk 8]
```diff
@@ -6,12 +6,11 @@
 package astmodel
 
 import (
-	"sort"
-	"strings"
-
 	"github.com/dave/dst"
 	"github.com/pkg/errors"
 	"k8s.io/klog/v2"
+	"sort"
```
Comment (matthchr) [inline hack/generator/pkg/astmodel/package_import_set.go:12]: minor: The reordering here seems incorrect?

I thought our pattern was:
```
stdlib

external

local
```

Which is what this originally had at the top?

File: hack/generator/pkg/astmodel/package_import_set.go [hunk 9]
```diff
@@ -269,7 +268,8 @@ func (set *PackageImportSet) ServiceNameForImport(imp PackageImport) string {
 //      for microsoft.storage/v20200101 extract "storagev20200101" and so on
 func (set *PackageImportSet) versionedNameForImport(imp PackageImport) string {
 	service := set.ServiceNameForImport(imp)
-	return service + imp.packageReference.PackageName()
+	version := imp.packageReference.PackageName()
+	return service + strings.Title(version)
```
Comment (matthchr) [inline hack/generator/pkg/astmodel/package_import_set.go:272]: Is it standard practice for package names in imports to be title-cased? According to https://blog.golang.org/package-names package names should be all lowercase all the time - I had assumed that extended to when you're naming a package something other than the default name during import too?

File: hack/generator/pkg/astmodel/package_reference.go [hunk 10]
```diff
@@ -36,3 +44,219 @@ func PackageAsLocalPackage(pkg PackageReference) (LocalPackageReference, error)
...
+
+		if v.left[v.leftIndex] == v.right[v.rightIndex] {
+			// Both runes the same, skip to the next one
+			v.leftIndex++
+			v.rightIndex++
+			continue
+		}
+
+		// Runes are different, make a decision
+		if v.left[v.leftIndex] < v.right[v.rightIndex] {
```
Comment (matthchr) [inline hack/generator/pkg/astmodel/package_reference.go:122]: you reuse `v.left[v.leftIndex]` and `v.right[v.rightIndex]` quite a lot here. Wondering if it makes sense to (after your bounds checks) do:
```
leftRune := v.left[v.leftIndex]
rightRune := v.right[v.rightIndex]
```

It doesn't necessarily save you a lot of text but I think it saves a good amount of cognitive load when parsing the statements.

File: hack/generator/pkg/astmodel/package_import_set_test.go [hunk 11]
```diff
@@ -297,24 +297,24 @@ func TestPackageImportSet_ResolveConflicts_GivenImplicityNamedConflicts_AssignsE
 			"Import conflicts with versioned resolution (i)",
 			createSet(emailTestRef, networkTestRef, emailTestAltRef, networkTestAltRef),
 			emailTestRef,
-			"emailv20180801",
+			"emailV20180801",
```
Comment (matthchr) [inline hack/generator/pkg/astmodel/package_import_set_test.go:300]: Same comment as above -- I think we're not expecting any uppercase characters here given it's a package name?

File: hack/generator/pkg/astmodel/package_reference.go [hunk 12]
```diff
@@ -36,3 +44,219 @@ func PackageAsLocalPackage(pkg PackageReference) (LocalPackageReference, error)
 	}
 	return LocalPackageReference{}, errors.Errorf("%q is not a local package", pkg.PackagePath())
 }
+
+func SortPackageReferencesByPathAndVersion(packages []PackageReference) {
```
Comment (matthchr) [inline hack/generator/pkg/astmodel/package_reference.go:48]: Is this for future use somewhere? I don't see it used in the main code at the moment, only in the test code.

File: hack/generator/pkg/astmodel/package_reference.go [hunk 13]
```diff
@@ -36,3 +44,219 @@ func PackageAsLocalPackage(pkg PackageReference) (LocalPackageReference, error)
...
+		}
+		compare := comparer.Compare()
+		return compare < 0
+	})
+}
+
+// versionComparer captures our state while doing an alphanumeric version comparision
+// We need separate indexes for each side because we're doing a numeric comparison, which will
+// compare "100" and "0100" as equal (leading zeros are not significant)
+type versionComparer struct {
```
Comment (matthchr) [inline hack/generator/pkg/astmodel/package_reference.go:62]: really minor: Given how much of this file is now the `versionComparer` wondering if we should put it into its own file?

Comment (matthchr) [reply hack/generator/pkg/astmodel/package_reference.go:62]: It really does only apply to this type though so I see why it's here...

File: hack/generator/pkg/astmodel/storage_package_reference.go [hunk 14]
```diff
@@ -29,8 +32,16 @@ func MakeStoragePackageReference(local LocalPackageReference) StoragePackageRefe
...
-func (spr StoragePackageReference) String() string {
-	return fmt.Sprintf("storage:%v", spr.PackagePath())
+func (s StoragePackageReference) String() string {
+	return fmt.Sprintf("storage:%v", s.PackagePath())
+}
+
+// IsPreview returns true if this package reference is a preview
+func (s StoragePackageReference) IsPreview() bool {
+	lc := strings.ToLower(s.version)
+	return strings.Contains(lc, "alpha") ||
```
Comment (matthchr) [inline hack/generator/pkg/astmodel/storage_package_reference.go:42]: Is there a way we could have this logic shared among the different package references? Possibly by defining a module-level function that does the check and then having each interface implementation call it? For example I see that you define `specialVersionIdentifiers` but then don't use it here -- a shared function could make use of that collection possibly and save a good bit of duplication?

Plus then there's only 1 place to change if we need to add/remove strings to it.

**Review by matthchr — APPROVED** [review]
LGTM -- had a few small comments though.

File: hack/generator/pkg/astmodel/package_reference.go (hunk 12 above)
Comment (theunrepentantgeek) [reply hack/generator/pkg/astmodel/package_reference.go:48]: Yeah, I need this when ordering storage packages to create the conversion graph.

File: hack/generator/pkg/astmodel/storage_package_reference.go (hunk 14 above)
Comment (theunrepentantgeek) [reply hack/generator/pkg/astmodel/storage_package_reference.go:42]: Good idea, done.

File: hack/generator/pkg/astmodel/package_import_set.go (hunk 8 above)
Comment (matthchr) [reply hack/generator/pkg/astmodel/package_import_set.go]: I think you can configure goland to do it for you -- I don't think gofmt does this one automatically but I could be wrong

File: hack/generator/pkg/astmodel/package_reference.go [hunk 15]
```diff
@@ -5,7 +5,13 @@
 
 package astmodel
 
-import "github.com/pkg/errors"
+import (
+	"sort"
+	strings "strings"
```
Comment (matthchr) [inline hack/generator/pkg/astmodel/package_reference.go:10]: minor: `strings "strings"` seems weird

File: hack/generator/pkg/astmodel/package_import_set.go (hunk 8 above)
Comment (theunrepentantgeek) [reply hack/generator/pkg/astmodel/package_import_set.go]: I already have GoLand configured to automatically apply **go fmt** when saving. Don't know what's going on here.

File: hack/generator/pkg/astmodel/package_reference.go (hunk 15 above)
Comment (theunrepentantgeek) [reply hack/generator/pkg/astmodel/package_reference.go]: Fixed.

**Pull request merged**